### ISSUE TITLE
Custom URL construction

### ISSUE THREAD

**Issue by mspae** [redux-json-api/redux-json-api#168]
Hello everybody! I would appreciate your feedback on this problem. I would be willing to work on a PR if we can agree on how this would be best handled.

A bit of backstory: I'm working on a high-level toolset to decouple Drupal with React. I'm using `redux-json-api` but I'm running into the following problem.

The JSON:API URL construction in Drupal is a little special (apparently because the data model is so complex) – Instead of the entity `type` property translating directly to the URL to access the JSON:API there it is namespaced.

So currently this entity:

```json
{
  "id": "xxxx",
  "type": "mytype",
  "attributes": {}
}
```

can be accessed with this URL: `/jsonapi/mytype/xxxx`

But with Drupal the entity looks like this:

```json
{
  "id": "xxxx",
  "type": "mytype--mybundle",
  "attributes": {}
}
```

And the URL is: `/jsonapi/mytype/mybundle/xxxx`

Since currently there is no transformation going on the request with this library currently resolves to: `/jsonapi/mytype--mybundle/xxxx`

This namespacing is not part of [the specification](http://jsonapi.org/recommendations/#urls) but I presume I won't be the only person with this problem. I propose an extension point to inject a URL construction function which receives the resource and returns the URL to use.

I think currently the easiest way would be to pass this function into all action creators that construct URLs in this way (`updateResource` and `deleteResource`)

```js
function constructApiEndpoint(resource) {
  return `${resource.type}/${resource.id}`
}
```

**Comment by mspae** [comment]
I've forked this and implemented it. Let me know if you want me to open a PR: https://github.com/mspae/redux-json-api/commit/24c4fd9c129f42b49d7cc0427a0e708549e06f53

**Comment by egeriis** [comment]
Hey @mspae!

Thank you for a thorough issue description. I think we should implement support for what you're asking for, but we may need to generalize it a bit more, so we can support more use cases. We've had a good chunk of requests for being able to plug into redux-json-api's functionality.

A few specific suggestions:

1. Implement a `beforeRequest` hook, that can be used to manipulate the request before it's made
1. Add a way to register a hook globally and per-request

At this point I wish we had implemented redux-json-api as a middleware, it would likely allow this problem to be solved a bit more elegantly, but here we are 😄

**Comment by mspae** [comment]
Agreed! 😃 

I'm just thinking if it would be possible to implement this with middlwares without breaking the publicly facing API. (Apart from the fact that you need to add the middleware of course)

I'm a bit new to redux middlewares so I'm not sure I've groked it. Hypothetically it would be possible to split the functionality into request trigger actions (the current public API actions) and do the actual data fetching in a middleware and return a promise, no? (I'm actually not sure if returning a promise would be possible with this architecture)

so instead of

1. updateResource()
2. API_WILL_UPDATE
3. apiRequest()
4. API_READ

it would be

1. updateResource()
2. API_WILL_UPDATE
3. API_REQUEST (action containing url, resource, axiosConfig)
4. custom middleware manipulations
5. library middleware triggers apiRequest()
6. API_READ
7. (potentially we could add another step here to allow extension middlewares to change resources before they enter the store)

… Without middleware we would need to provide some form of instance of redux-json-api to inject the hook(s) into. And to extend/override those hook(s) we would pass in a hooks object as second parameter to the api action creator functions (kind of like my approach but with different types of hooks)

I'm willing to work on this problem. I believe the first one would indeed be a much nicer solution but I think it currently hinges on the question whether we can get the middleware to return promises on the original function. What do you think?

**Comment by egeriis** [comment]
I think something like this would work. One of the main advantages is easy global configuration and the ability to specify which state key you would like redux-json-api mounted on.

```js
import { configure } from 'redux-json-api';

const reduxJsonApi = configure(stateKey = 'api', options = {
  baseUrl: 'https://api.example.com',
  hooks: {
    beforeRequest: (request) => console.log(request);
  }
});

configureStore([thunk, reduxJsonApi]);
```

Then the middleware would just have to register the redux-json-api reducer. E.g. using [`replaceReducer`](https://github.com/reduxjs/redux/blob/master/docs/api/Store.md#replacereducernextreducer)

**Comment by mspae** [comment]
So implement this using a middleware for the initialization of redux-json-api but not for the actual asynchronous API requests?

Where would the `hooks` be cached, in some global reduxJsonApi instance?

**Comment by egeriis** [comment]
> So implement this using a middleware for the initialization of redux-json-api but not for the actual asynchronous API requests?

Yeah, so we don't need to change how actions are dispatched. I don't think that's necessarily.

> Where would the `hooks` be cached, in some global reduxJsonApi instance?

Yeah, should be fine to just have a var in [jsonapi.js](https://github.com/redux-json-api/redux-json-api/blob/master/src/jsonapi.js) where we store these configuration details.

**Comment by mspae** [comment]
Okay so I've been thinking. Because I want to make this as extendible as possible within this framework I'm building and middlewares are of course the best way to do that:

We add the reducer via middleware with `replaceReducer` like you said. In addition we create an `EventEmitter` and cache that. Then everytime you call for instance `updateResource` we go through the whole asynchronous middleware process (Like I outlined in https://github.com/redux-json-api/redux-json-api/issues/168#issuecomment-424676791), returning a Promise from that action creator function which is resolved once the `EventEmitter` fires a "API Request resolved" event with the "original payload" matching the payload we called `updateResource` with.  The real payload can manipulated by middlewares along the way, the original payload stays the same so our `updateResource` knows when it can resolve. The same process applies to rejecting the promise of course.

This way we don't change the public facing API and still have a fully customisable middleware approach.

Of course there is additional indirection and code complexity but as far as I can tell this will work just fine.

If you're okay with this I would be willing to put together a basic implementatin we can discuss further within the next week.

Cheers!

**Comment by egeriis** [comment]
If I understand your correctly, the middleware is responsible for global configuration, while the EventEmitter enables hooking into the various API requests. Sounds good to me 🙂 

I think everything should be configurable "by instance", i.e. with this new approach you could potentially have two redux-json-api reducers with different configurations, two "instances".

**Comment by mspae** [comment]
The middleware is responsible for manipulating the payload, performing the request and then passing the data to the reducer (this way people can hook into any of these steps along the way, e.g. to change the endpoint URL), the EventEmitter is responsible for resolving/rejecting the original action creator Promises so the original API doesn't break (and so we don't loose this very useful feature that the action creator returns a promise)

This is absolutely compatible with a per-instance approach. (And I think that is a good idea)

We need to think how we will bind the action creators to the right instance. Maybe this would be a good use case for an (optional) second action creator parameter which describes for which state slice/instance the action should be performed. I think this is probably the only solution if we want to prevent breakage. (The other option would be to create the action creators in the `configure` function you outlined in https://github.com/redux-json-api/redux-json-api/issues/168#issuecomment-424763464 but I think this is not doable without breaking the current API)

**Comment by mspae** [comment]
Right. Scratch that. I think I've been overcomplicating things.

New idea: Everything stays the same, the API action functions receive an optional second argument (`options`) like some already do. Within this you can add your per-request custom `axiosConfig`, `stateKey`(default: `api`) etc.

Before the options are actually used in the request it is set to to the store in a `requestConfig` property. It is then immediately read out again with `getState()` and used. This way middleware can adjust the payload, endpoint etc. and you can set per-request-level options as well (with the function parameters).

This enables a per-instance approach because the request type, the state slice etc. are all part of the action and can be queried in any middleware wishing to customise.

I think the configuration middleware we talked about should be very simple (something like this https://github.com/redux-json-api/react-redux-json-api-example/blob/master/src/store.js#L8) – and maybe it should even just be shown in the docs as a "how to customise"-starter, not as an exposed part of the library itself.

I already have a pretty complete setup of this, but tests are still failing. Hopefully I will have something to show later in the week.
